soc/arm/silabs_exx32/common/soc_power.c

```diff
@@ -63,3 +157,136 @@ __weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+					   lf_ticks_to_sleep,
+					   sli_schedule_wakeup_timer_expire_handler,
+					   0,
+					   0,
+					   0);
+	__ASSERT(status == SL_STATUS_OK, "Failed to create timer");
+
+}
+
+static void sli_schedule_wakeup_timer_expire_handler(sl_sleeptimer_timer_handle_t *handle,
+						     void *data)
+{
+	(void)handle;
+	(void)data;
+
+	/* This handler is just to wakeup the device from the low energy mode */
+}
+
+/* Function to check any ready threads are scheduled */
+bool ConfirmSleepModeStatus(void)
+{
+	bool ret = true;
+
+	if (!z_is_idle_thread_object(_kernel.ready_q.cache)) {
+		/* Issue: sysTick Counter is not incrementing if deep sleep bit is set
+		 * WorkAround: Clearing the Deep Sleep Bit , incase not cleared by
+		 * sl_power_manager_sleep
+		 */
+		SCB->SCR &= ~SCB_SCR_SLEEPDEEP_Msk;
+
+		isContextSwitchRequired = true;
+
+		ret = false;
+	}
+	return ret;
+}
+
+/* Function called by power manager to determine if system can go back to sleep
+ * after a wakeup
+ */
+bool sl_power_manager_sleep_on_isr_exit(void)
+{
+	uint32_t slept_lf_ticks;
+	uint32_t slept_os_ticks;
+
+	/* Determine how long we slept. */
+	slept_lf_ticks = sl_sleeptimer_get_tick_count() - last_update_lftick;
+	slept_os_ticks = (slept_lf_ticks * CONFIG_SYS_CLOCK_TICKS_PER_SEC) / sleeptimer_freq;
+	last_update_lftick += slept_lf_ticks;
+
+	/* Notify RTOS of how long we slept. */
+	if ((total_slept_lf_ticks + slept_lf_ticks) < expected_sleep_lf_ticks) {
+		total_slept_lf_ticks += slept_lf_ticks;
+	} else {
+		slept_os_ticks = ((expected_sleep_lf_ticks - total_slept_lf_ticks) *
+				   CONFIG_SYS_CLOCK_TICKS_PER_SEC) / sleeptimer_freq;
+		total_slept_lf_ticks = expected_sleep_lf_ticks;
+	}
+
+	if (SCB->SCR & SCB_SCR_SLEEPDEEP_Msk) {
+		sys_clock_announce(slept_os_ticks);
+	}
+
+	/* Have we slept enough ? */
+	if (total_slept_lf_ticks >= expected_sleep_lf_ticks) {
+		return false;
+	}
+	/*  Check if we can sleep again */
+	return (ConfirmSleepModeStatus() != false);
+}
+
+/* Function called by power manager to determine if system can go to sleep
+ * or not
+ */
+bool sl_power_manager_is_ok_to_sleep(void)
+{
+	return (ConfirmSleepModeStatus() != false);
+}
+
+#ifndef CONFIG_PM_POLICY_DEFAULT
+/* CONFIG_PM_POLICY_CUSTOM allows to define application specific power policy
+ * checking whether the ticks is meeting the minimum criteria to enter
+ * into SOC specific power policy or not
+ */
+struct pm_state_info *pm_policy_next_state(uint8_t cpu, int32_t ticks)
+{
+	int i;
+
+	for (i = ARRAY_SIZE(pm_min_residency) - 1; i >= 0; i--) {
+
+		if ((ticks == K_TICKS_FOREVER) ||
+		    (ticks >= k_us_to_ticks_ceil32(
+			     pm_min_residency[i].min_residency_us))) {
+			LOG_DBG("Selected power state %d "
+				"(ticks: %d, min_residency: %u)",
+				pm_min_residency[i].state, ticks,
+				pm_min_residency[i].min_residency_us);
+
+			/* Busy waiting for 1 tick to flush UART buffer */
+			k_busy_wait(k_ticks_to_us_floor32(1));
+			IdleTicks = ticks - 1;
```

Comment:
what's going on with UART here?

Reply:
It looks like the UART peripheral ends in some kind of stale state if we don't shorten the sleep time. This does look like some kind of race condition, I'm looking into this.

Reply:
You need to lock PM states if UART prevents from using them. Some UART drivers do this.

Reply:
That was exactly the case here. Thank you for your suggestion, it helped me a lot!